components/file_system/src/iosnoop/biosnoop.c

```diff
@@ -0,0 +1,204 @@
+BPF_HISTOGRAM(load_balance_read_latency, int, 25);
+BPF_HISTOGRAM(import_read_latency, int, 25);
+BPF_HISTOGRAM(export_read_latency, int, 25);
+
+BPF_HISTOGRAM(other_write_latency, int, 25);
+BPF_HISTOGRAM(read_write_latency, int, 25);
+BPF_HISTOGRAM(write_write_latency, int, 25);
+BPF_HISTOGRAM(coprocessor_write_latency, int, 25);
+BPF_HISTOGRAM(flush_write_latency, int, 25);
+BPF_HISTOGRAM(compaction_write_latency, int, 25);
+BPF_HISTOGRAM(replication_write_latency, int, 25);
+BPF_HISTOGRAM(load_balance_write_latency, int, 25);
+BPF_HISTOGRAM(import_write_latency, int, 25);
+BPF_HISTOGRAM(export_write_latency, int, 25);
+
+// cache PID by req
+int trace_req_start(struct pt_regs *ctx, struct request *req) {
+  // TODO: Using bpf_get_ns_current_pid_tgid of newer kernel to get pid under
+  // namespace
+  //
+  // struct bpf_pidns_info ns = {};
+  // if (!bpf_get_ns_current_pid_tgid(##DEV##, ##INO##, &ns, sizeof(struct
+  // bpf_pidns_info))) {
+  //     return 0;
+  // }
+  // u32 pid = ns.pid;
+  // u32 tgid = ns.tgid;
+
+  u64 id = bpf_get_current_pid_tgid();
+  u32 pid = id & (((u64)1 << 32) - 1);
+  u32 tgid = id >> 32;
+  if (tgid != ##TGID##) {
+    return 0;
+  }
+
+  io_type **type_ptr = type_by_pid.lookup(&pid);
+  if (type_ptr == 0)
+    return 0;
+  struct info_t info;
+  __builtin_memset(&info, 0, sizeof(info)); // ensure padding is always zeroed,
+                                            // otherwise verifier will complain.
+  int err = bpf_probe_read(&info.type, sizeof(io_type), (void *)*type_ptr);
+  if (err != 0) {
+    info.type = Other;
+  }
+  info.start_ts = bpf_ktime_get_ns();
+
+  info_by_req.update(&req, &info);
+  return 0;
+}
+
+// trace_req_completion may be called in interrput context. In that case,
+// `bpf_get_current_pid_tgid` and `bpf_probe_read` can not work as expected.
+// So caching type in `trace_pid_start` which wouldn't be called in interrupt
+// context and query the type by req in `info_by_req`.
+int trace_req_completion(struct pt_regs *ctx, struct request *req) {
+  struct info_t *info = info_by_req.lookup(&req);
+  if (info == 0)
+    return 0;
+  /*
+   * The following deals with a kernel version change (in mainline 4.7, although
+   * it may be backported to earlier kernels) with how block request write flags
+   * are tested. We handle both pre- and post-change versions here. Please avoid
+   * kernel version tests like this as much as possible: they inflate the code,
+   * test, and maintenance burden.
+   */
+  u64 rwflag;
+#ifdef REQ_WRITE
+  rwflag = !!(req->cmd_flags & REQ_WRITE);
+#elif defined(REQ_OP_SHIFT)
+  rwflag = !!((req->cmd_flags >> REQ_OP_SHIFT) == REQ_OP_WRITE);
+#else
+  rwflag = !!((req->cmd_flags & REQ_OP_MASK) == REQ_OP_WRITE);
+#endif
+  io_type type = info->type;
+  struct stats_t zero = {}, *val;
+  val = stats_by_type.lookup_or_init(&type, &zero);
+  if (rwflag == 1) {
+    __sync_fetch_and_add(&val->write, req->__data_len);
```

Comment:
shall also use atomic when flushing metrics.

Reply:
when flushing metrics, it gets the map value as a whole, so it can't use atomic to get it.

Reply:
IIUC the map values are used as raw memory buffer, and we can cast each cell as atomic int and operate on them. Not sure how to write it in rust though.